File: experiments/old/test_ragft_forward_inverse.py
import numpy as np
from ply import ply_read8i, ply_write
from color_conversion import RGBtoYUV, YUVtoRGB
from ragft import Region_Adaptive_GFT, iRegion_Adaptive_GFT

# Load point cloud data
filename = 'longdress_vox10_1051.ply'
V, Crgb, J = ply_read8i(filename)

# Convert RGB to YUV
C = RGBtoYUV(Crgb)

# Parameters for RA-GFT
bsize = [2, 2, 2, 2, 2, 2, 2, 2, 2, 2]
param = {'V': V, 'J': J, 'bsize': bsize, 'isMultiLevel': 0}

# Perform RA-GFT encoding
step = 64
Coeff, Gfreq, weights = Region_Adaptive_GFT(C, param)

# Quantize coefficients
Coeff_quant = np.round(Coeff / step) * step

# Perform inverse RA-GFT
start_indices, end_indices, V_MR, Crec = iRegion_Adaptive_GFT(Coeff_quant, param)

# Convert reconstructed YUV to RGB
Crgb_rec = YUVtoRGB(Crec)

# Calculate PSNR for luminance channel
Y = Coeff[:, 0]
psnr_Y = -10 * np.log10(np.linalg.norm(Y - Coeff_quant[:, 0])**2 / (N * 255**2))

# Write original and reconstructed point clouds to PLY files
ply_write('PC_original.ply', V, Crgb, [])
ply_write('PC_coded.ply', V, Crgb_rec, [])

File: matlab_code/test_ragft_forward_inverse.m
% TODO: HACER ESTEEEEE PARA MI TRANSFORMADA

% Authors - Eduardo Pavez <eduardo.pavez.carvelli@gmail.com, pavezcar@usc.edu>
% Copyright Eduardo Pavez,  University of Southern California
%Los Angeles, USA, 05/30/2020
% E. Pavez, B. Girault, A. Ortega, and P. A. Chou. 
%"Region adaptive graph Fourier transform for 3D point clouds". 
%IEEE International Conference on Image Processing (ICIP), 2020
%https://arxiv.org/abs/2003.01866
%% script to test RA-GFT and its inverse
%encoder RA-GFT
clear;
filename = 'longdress_vox10_1051.ply';
[V,Crgb,J] = ply_read8i(filename);              % V lista de puntos, Crgb lista de colores, J resolución de voxel 2^10 se dividio 10 veces en cada cubo ahí un punto en cada voxel
N = size(V,1);
C = RGBtoYUV(Crgb); %transform to YUV

%%
bsize=[ 2 2 2 2 2 2  2 2 2 2]; % Tamaño de cada nivel en caso multilevel = 0. Tamaño 2 en que se dividen los bloques a lo más 2^3 8 puntos máximo. (8, 16)
param.V=V;
param.J=J;
param.bsize = bsize;
param.isMultiLevel=0;           % Solo una vez, setearlo a cero
tic;

step = 64;
%C = ones(N,1);
[Coeff, Gfreq, weights]  = Region_Adaptive_GFT( C, param ); % param posee la información definida
toc;
Y = Coeff(:,1);
Coeff_quant = round(Coeff/step)*step;
%%
tic;
[ start_indices, end_indices, V_MR, Crec ] = iRegion_Adaptive_GFT( Coeff_quant, param );
toc;

Crgb_rec = double(YUVtoRGB(Crec));

psnr_Y = -10*log10(norm(Y - Coeff_quant(:,1))^2/(N*255^2));

%%
 ply_write('PC_original.ply',V,Crgb,[]);
 ply_write('PC_coded.ply',V,Crgb_rec,[]);

